ENIGMAsystem/SHELL/Platforms/General/UNIXfilemanip.cpp

```diff
@@ -20,48 +20,626 @@
+        while (*str2.rbegin() == '/')
+        {
+            str2.erase(str2.size()-1);
+        }
+    }
+
+    size_t slash1;
+    size_t slash2;
+    std::string dir1(str1);
+    std::string dir2(str2);
+
+    if (dir1.find('/') != std::string::npos)
+        slash1 = dir1.find_last_of("/");
+    else
+    {
+        dir1 = std::string("");
+        slash1 = 0;
+    }
+
+    if (dir2.find('/') != std::string::npos)
+        slash2 = dir2.find_last_of("/");
+    else
+    {
+        dir2 = std::string("");
+        slash2 = 0;
+    }
+
+    int proceed1 = (str1 != str2.substr(0,str1.length()));
+    int proceed2 = (dir1.substr(0,slash1) == dir2.substr(0,slash2));
+
+    std::string prestr1 = str1;
+    std::string prestr2 = str2;
+    std::string poststr1 = str1;
+    std::string poststr2 = str2;
+
+    std::string pardir1(dirname((char *)str1.c_str()));
+    if (!pardir1.empty())
+    {
+        while (*pardir1.rbegin() == '/')
+        {
+            pardir1.erase(pardir1.size()-1);
+        }
+    }
+    struct stat pexist1;
+    if (stat((char *)pardir1.c_str(),&pexist1) == 0 &&
+        S_ISDIR(pexist1.st_mode))
+    {
+        char actpath1[PATH_MAX+1];
+        char *ptr1 = realpath((char *)pardir1.c_str(),actpath1);
+        if (ptr1 != NULL)
+        {
+            std::string pardir2(dirname((char *)str2.c_str()));
+            if (!pardir2.empty())
+            {
+                while (*pardir2.rbegin() == '/')
+                {
+                    pardir2.erase(pardir2.size()-1);
+                }
+            }
+            struct stat pexist2;
+            if (stat((char *)pardir2.c_str(),&pexist2) == 0 &&
+                S_ISDIR(pexist2.st_mode))
+            {
+                char actpath2[PATH_MAX+1];
+                char *ptr2 = realpath((char *)pardir2.c_str(),actpath2);
+                if (ptr2 != NULL)
+                {
+                    std::string parbas1(std::string("/")+basename((char *)poststr1.c_str()));
+                    std::string parbas2(std::string("/")+basename((char *)poststr2.c_str()));
+
+                    proceed1 = (std::string(actpath1+parbas1) != std::string(actpath2+parbas2).substr(0,std::string(actpath1+parbas1).length()));
+                    proceed2 = (std::string(actpath1) == std::string(actpath2));
+                }
+            }
+        }
+    }
+
+    str1 = prestr1;
+    str2 = prestr2;
+
+    if (proceed1 || (proceed2 && basename((char *)str1.c_str()) != basename((char *)str2.c_str())))
+    {
+        struct stat sb1;
+        struct stat sb2;
+        if (stat((char *)str1.c_str(),&sb1) == 0 &&
+            S_ISDIR(sb1.st_mode) &&
+            stat((char *)str2.c_str(),&sb2) != 0)
+        {
+            return (rename((char *)str1.c_str(),(char *)str2.c_str()) == 0);
+        }
+    }
+
+    return 0;
 }
 
-int directory_delete(string dname) {
-  return rmdir(dname.c_str());
+int directory_copy(std::string dname,std::string newname)
+{
+    std::string str1(dname);
```

Comment:
More helper functions, more FTW. This entire function should be a few lines in length. Something like this:

```C++
static std::string dst_root;
return ftw(src_root, [](const char* src_path, const struct stat* st, int typeflag) {
    std::string dst_path = dst_root + src_path;
    if (typeflag == FTW_D) return mkdir(dst_path.c_str(), st->st_mode);
    return !file_copy(src_path, dst_path);
}, 5);
```

If you need a special `file_copy` to handle hard links, that's fine, but I believe it would behoove our regular `file_copy` to handle links.